pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,500 @@
+//
+// See docs of NeoGo configuration for some details.
+type Config struct {
+	// Writer of the Blockchain's logs and internal errors.
+	//
+	// Optional: by default, Blockchain doesn't write logs.
+	Logger *zap.Logger
+
+	// Application level error listener. Blockchain writes any internal error to the
+	// channel. Error returns of functions (e.g. New) are not pushed. The channel
+	// should be regularly checked in order to prevent blocking. Any pop-up error
+	// does not allow the Blockchain to fully work, therefore, if it is detected,
+	// the blockchain should be stopped.
+	//
+	// Required.
+	ErrorListener chan<- error
+
+	// Identifier of the Neo network.
+	//
+	// Required.
+	NetworkMagic netmode.Magic
+
+	// Initial committee staff.
+	//
+	// Required.
+	Committee keys.PublicKeys
+
+	// Time period (approximate) between two adjacent blocks.
+	//
+	// Optional: defaults to 15s. Must not be negative.
+	BlockInterval time.Duration
+
+	// Network addresses to listen Neo RPC on. Each element must be a valid TCP
+	// address in 'host:port' format.
+	//
+	// Optional: by default, Neo RPC is not served.
+	RPCListenAddresses []string
+
+	// Length of the chain accessible to smart contracts.
+	//
+	// Optional: defaults to 2102400.
+	TraceableChainLength uint32
+
+	// Maps hard-fork's name to the appearance chain height.
+	//
+	// Optional: by default, each known hard-fork is applied from the zero
+	// blockchain height.
+	HardForks map[string]uint32
+
+	// List of nodes' addresses to communicate with over Neo P2P protocol in
+	// 'host:port' format.
+	//
+	// Optional: by default, node runs as standalone.
+	SeedNodes []string
+
+	// Disable Notary service.
+	//
+	// Optional.
+	NotaryDisabled bool
+
+	// P2P settings.
+	//
+	// Required.
+	P2P P2PConfig
+
+	// Storage configuration. Must be set using one of constructors like BoltDB.
+	//
+	// Required.
+	Storage StorageConfig
+
+	// NEO wallet of the node. The wallet is used by Consensus and Notary services.
+	// Corresponding private key be accessed via Blockchain.LocalKey.
+	//
+	// Required.
+	Wallet config.Wallet
+
+	// Maps chain height to number of consensus nodes.
+	//
+	// Optional: by default Committee size is used. Each value must not be greater
+	// than math.MaxInt32.
+	ValidatorsHistory map[uint32]uint32
+
+	// Chronology of native contracts updates. Maps name of Neo native contract to
+	// chain heights.
+	//
+	// Optional: by default, all native contracts are active from genesis block.
+	// Keys must be valid native names. Values must not be empty.
+	NativeActivations map[string][]uint32
+}
+
+// New returns new Blockchain configured by the specified Config. New panics if
+// any required Config field is zero or unset. Resulting Blockchain is ready to
+// run. Launched Blockchain should be finally stopped.
+func New(cfg Config) (*Blockchain, error) {
+	switch {
+	case cfg.Storage.typ == "":
+		panic("uninitialized storage config")
```

Comment:
Why do we panic here and below? Why not to provide some logger and do `log.Fatal(...)` or just return an error? Is there any recover from this panic? It's not very pleasent when the user sees panic on the screen.

Reply:
Because that's developer failure, not user. All requirements are strict and stated in the documentation, they are checked at the upper level. The package is internal an not intended to be widely used. For example, stdlib panics on input sometimes and you should care about it.

I don't mind to return an error, but at the end of the day this is still a programmer's mistake.

@roman-khimov thoughts?

Reply:
This particular check is internal, not related to user data (which is checked in `parseBlockchainConfig`), so yes, it can be a panic here.